File: app/src/main/java/jp/hazuki/yuzubrowser/action/item/VibrationSingleAction.java
package jp.hazuki.yuzubrowser.action.item;

import android.content.DialogInterface;
import android.os.Parcel;
import android.os.Parcelable;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.core.JsonToken;

import java.io.IOException;

import jp.hazuki.yuzubrowser.R;
import jp.hazuki.yuzubrowser.action.SingleAction;
import jp.hazuki.yuzubrowser.action.view.ActionActivity;
import jp.hazuki.yuzubrowser.utils.app.StartActivityInfo;
import jp.hazuki.yuzubrowser.utils.view.SeekBarDialog;

public class VibrationSingleAction extends SingleAction implements Parcelable {
    private static final String FIELD_NAME_TIME = "0";
    private int mTime = 100;

    public VibrationSingleAction(int id, JsonParser parser) throws IOException {
        super(id);

        if (parser != null) {
            if (parser.nextToken() != JsonToken.START_OBJECT) return;
            while (parser.nextToken() != JsonToken.END_OBJECT) {
                if (parser.getCurrentToken() != JsonToken.FIELD_NAME) return;
                if (!FIELD_NAME_TIME.equals(parser.getCurrentName())) {
                    parser.skipChildren();
                    continue;
                }
                if (parser.nextToken() != JsonToken.VALUE_NUMBER_INT) return;
                mTime = parser.getIntValue();
            }
        }
    }

    @Override
    public void writeIdAndData(JsonGenerator generator) throws IOException {
        generator.writeNumber(id);
        generator.writeStartObject();
        generator.writeNumberField(FIELD_NAME_TIME, mTime);
        generator.writeEndObject();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(id);
        dest.writeInt(mTime);
    }

    protected VibrationSingleAction(Parcel source) {
        super(source.readInt());
        mTime = source.readInt();
    }

    public static final Creator<VibrationSingleAction> CREATOR = new Creator<VibrationSingleAction>() {
        @Override
        public VibrationSingleAction createFromParcel(Parcel source) {
            return new VibrationSingleAction(source);
        }

        @Override
        public VibrationSingleAction[] newArray(int size) {
            return new VibrationSingleAction[size];
        }
    };

    @Override
    public StartActivityInfo showSubPreference(ActionActivity context) {
        new SeekBarDialog(context)
                .setTitle(R.string.action_vibration_setting)
                .setPositiveButton(android.R.string.ok, new SeekBarDialog.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which, int value) {
                        mTime = value;
                    }
                })
                .setSeekMin(1)
                .setSeekMax(3000)
                .setValue(mTime)
                .setNegativeButton(android.R.string.cancel, null)
                .show();

        return null;
    }

    public int getTime() {
        return mTime;
    }
}

File: legacy/src/main/java/jp/hazuki/yuzubrowser/legacy/action/item/VibrationSingleAction.kt
/*
 * Copyright (C) 2017-2019 Hazuki
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.legacy.action.item

import android.os.Parcel
import android.os.Parcelable
import com.squareup.moshi.JsonReader
import com.squareup.moshi.JsonWriter
import jp.hazuki.yuzubrowser.legacy.R
import jp.hazuki.yuzubrowser.legacy.action.SingleAction
import jp.hazuki.yuzubrowser.legacy.action.view.ActionActivity
import jp.hazuki.yuzubrowser.ui.app.StartActivityInfo
import jp.hazuki.yuzubrowser.ui.dialog.SeekBarDialog
import java.io.IOException

class VibrationSingleAction : SingleAction, Parcelable {
    var time = 100
        private set

    @Throws(IOException::class)
    constructor(id: Int, reader: JsonReader?) : super(id) {
        if (reader != null) {
            if (reader.peek() != JsonReader.Token.BEGIN_OBJECT) return
            reader.beginObject()
            while (reader.hasNext()) {
                if (reader.peek() != JsonReader.Token.NAME) return
                when (reader.nextName()) {
                    FIELD_NAME_TIME -> {
                        if (reader.peek() != JsonReader.Token.NUMBER) return
                        time = reader.nextInt()
                    }
                    else -> reader.skipValue()
                }
            }
            reader.endObject()
        }
    }

    @Throws(IOException::class)
    override fun writeIdAndData(writer: JsonWriter) {
        writer.value(id)
        writer.beginObject()
        writer.name(FIELD_NAME_TIME)
        writer.value(time)
        writer.endObject()
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeInt(id)
        dest.writeInt(time)
    }

    private constructor(source: Parcel) : super(source.readInt()) {
        time = source.readInt()
    }

    override fun showSubPreference(context: ActionActivity): StartActivityInfo? {
        SeekBarDialog(context)
                .setTitle(R.string.action_vibration_setting)
                .setPositiveButton(android.R.string.ok) { _, _, value -> time = value }
                .setSeekMin(1)
                .setSeekMax(3000)
                .setValue(time)
                .setNegativeButton(android.R.string.cancel, null)
                .show()

        return null
    }

    companion object {
        private const val FIELD_NAME_TIME = "0"

        @JvmField
        val CREATOR: Parcelable.Creator<VibrationSingleAction> = object : Parcelable.Creator<VibrationSingleAction> {
            override fun createFromParcel(source: Parcel): VibrationSingleAction {
                return VibrationSingleAction(source)
            }

            override fun newArray(size: Int): Array<VibrationSingleAction?> {
                return arrayOfNulls(size)
            }
        }
    }
}
